packages/test/src/Happy/Test.hs

```diff
@@ -0,0 +1,86 @@
+module Happy.Test(test, TestSetup(..), defaultTestFiles, attributeGrammarTestFiles, defaultArguments) where
+
+import Happy.Test.Shell
+import System.IO
+import System.FilePath
+import Control.Exception
+import System.Directory
+import System.Exit
+import Paths_happy_test
+
+data TestSetup = TestSetup {
+  happyExec :: String,      -- name of the happy exeuctable which shall be tested.
+  defaultTests :: [String], -- standard tests from happy-test package that should be performed. these are in this package's data-dir
+  customTests :: [String],  -- custom tests from the calling package that should be performed. these are in the calling package's data-dir
+  customDataDir :: String,  -- data-dir of the calling package. all tests are compiled and executed in their respective directory.
+  allArguments :: [String], -- all different testable argument combinations for happy, as strings
+  stopOnFailure :: Bool     -- continue with remaining tests after an error has occurred?
+}
+
+test :: TestSetup -> IO a
+test setup = do
+  hSetBuffering stdout NoBuffering -- required for cabal test
+  defaultDir <- getDataDir
+  let files = zip (repeat defaultDir) (defaultTests setup) ++
+                zip (repeat (customDataDir setup)) (customTests setup)              -- (dir, file.ly)
+  let tests = [(dir, file, arg) | (dir, file) <- files, arg <- allArguments setup]  -- (dir, file.ly, -ag)
+  result <- testCons tests (happyExec setup) (stopOnFailure setup)
+  if result then exitSuccess else exitFailure
+
+testCons :: [(String, String, String)] -> String -> Bool -> IO Bool
+testCons [] _ _ = return True
+testCons ((dir, file, args):rest) happy stopOnFail = do
+  result <- runSingleTest happy args dir file
+  if result then testCons rest happy stopOnFail
+    else if stopOnFail
+      then return False
+      else do _ <- testCons rest happy stopOnFail; return False
```

Comment:
The "Cons" suffix is a bit misleading. In Haskell, "cons" usually has something to with the list constructor `(:)`. The terminology is borrowed from LISP, where this constructor is even called `cons`.